sdr_airspy.c

```diff
@@ -0,0 +1,467 @@
+        }
+
+        if (AIRSPY.enable_lna_agc) {
+            status = airspy_set_lna_agc(AIRSPY.dev, 1);
+            AIRSPY_STATUS(status, "airspy_set_lna_agc failed");
+            fprintf(stderr, "LNA Gain: AGC, ");
+        } 
+        else {
+            status = airspy_set_lna_gain(AIRSPY.dev, AIRSPY.lna_gain);
+            AIRSPY_STATUS(status, "airspy_set_lna_gain failed");
+            fprintf(stderr, "LNA Gain: %i, ", AIRSPY.lna_gain);
+        }
+
+        status = airspy_set_vga_gain(AIRSPY.dev, AIRSPY.vga_gain);
+        AIRSPY_STATUS(status, "airspy_set_vga_gain failed");
+        fprintf(stderr, "VGA Gain: %i, ", AIRSPY.vga_gain);
+    }
+
+    if (AIRSPY.enable_airspy_biast) {
+        status = airspy_set_rf_bias(AIRSPY.dev, 1);
+        AIRSPY_STATUS(status, "airspy_set_rf_bias (on) failed");
+        fprintf(stderr, "Bias-t: On\n");
+    }
+    else {
+        status = airspy_set_rf_bias(AIRSPY.dev, 0);
+        AIRSPY_STATUS(status, "airspy_set_rf_bias (off) failed");
+        fprintf(stderr, "Bias-t: Off\n");
+    }
+
+    AIRSPY.converter = init_converter(INPUT_UC8,
+                                      Modes.sample_rate,
+                                      Modes.dc_filter,
+                                      &AIRSPY.converter_state);
+
+    if (!AIRSPY.converter) {
+        fprintf(stderr, "airspy: can't initialize sample converter\n");
+        airspyClose();
+        return false;
+    }
+
+    fprintf(stderr, "Airspy successfully initialized\n");
+
+    return true;
+}
+
+static struct timespec airspy_thread_cpu;
+
+int airspyCallback(airspy_transfer *transfer) {
+    int16_t *inptr = (int16_t *)transfer->samples;
+    int16_t *outptr = (int16_t *)AIRSPY.airspy_scratch;
+    size_t i_len = transfer->sample_count;
+    size_t i, i_done, o_done;
+    struct mag_buf *outbuf;
+    struct mag_buf *lastbuf;
+    uint32_t slen;
+    unsigned next_free_buffer;
+    unsigned free_bufs;
+    unsigned block_duration;
+    static int dropping = 0;
+    static uint64_t sampleCounter = 0;
+
+    // Lock the data buffer variables before accessing them
+    pthread_mutex_lock(&Modes.data_mutex);
+    if (Modes.exit) {
+        airspy_stop_rx(AIRSPY.dev);
+        return 0;
+    }
+
+    next_free_buffer = (Modes.first_free_buffer + 1) % MODES_MAG_BUFFERS;
+    outbuf = &Modes.mag_buffers[Modes.first_free_buffer];
+    lastbuf = &Modes.mag_buffers[(Modes.first_free_buffer + MODES_MAG_BUFFERS - 1) % MODES_MAG_BUFFERS];
+    free_bufs = (Modes.first_filled_buffer - next_free_buffer + MODES_MAG_BUFFERS) % MODES_MAG_BUFFERS;
+
+    // Give the demodulater what it expects
+    soxr_process(AIRSPY.resampler, inptr, i_len, &i_done, outptr, MODES_RTL_BUF_SIZE, &o_done);
+    for (i = 0; i < o_done; i++)
+        AIRSPY.airspy_bytes[i] = (int8_t)(outptr[i] >> 4) + 127;
+
+    // The resampler often produces an odd sample count, no real value in tracking how often
+    if (o_done & 1)
+        o_done -= 1;
```

Comment:
So, no surprise... the airspy is dropping a load of samples.

What is a bit of a surprise to me is that I ran a quick test that didn't run the resampler and instead just passed the buffer along just to see what effect it would have on utilization (obviously aware nothing would be decoded). I thought that the resampler was the primary utilization issue, but to the thread was still hammered at 100%. No doubt easy to explain, your insight would be appreciated.

And now a new problem... faup1090 is failing to connect to 30005. It's repeatedly trying to start with '/usr/lib/piaware/helpers/faup1090 --net-bo-ipaddr {} --net-bo-port 30005 --stdout --lat...' and failing with ' failed to connect to :30005 (is dump1090 running?): can't resolve : Name or service not known'. dump1090-fa is indeed running. mlat-client is also giving ' Connection to :30005 failed: [Errno -2] Name or service not known'. The /etc/default/dump1090-fa file looks fine.